File: daz-ts/src/bone.ts
import type { DazClient } from "./client.js";
import { DazNode, type NodeIdentifier } from "./node.js";
import { ScriptBuilder } from "./scriptBuilder.js";

/**
 * Proxy for a `DzBone` (a single joint within a {@link DazSkeleton}).
 * Extends `DazNode` with bone-specific rotation helpers.
 */
export class DazBone extends DazNode {
  /**
   * Construct from a standard scene node lookup by identifier.
   */
  constructor(client: DazClient, identifier: NodeIdentifier);
  /**
   * Construct from a pre-built skeleton-relative locator (used by {@link fromLocator}).
   * The identifier's value provides the bone's name.
   */
  constructor(client: DazClient, identifier: NodeIdentifier, locator: string);
  constructor(client: DazClient, identifier: NodeIdentifier, locator?: string) {
    if (locator !== undefined) {
      super(client, identifier, locator);
    } else {
      super(client, identifier);
    }
  }

  /** Construct a `DazBone` from a pre-built skeleton-relative locator (see `DazSkeleton`'s bone lookup). */
  static fromLocator(client: DazClient, locator: string, name: string): DazBone {
    return new DazBone(client, { value: name, kind: "name" }, locator);
  }

  private nb(body: string): string {
    return ScriptBuilder.nodeBodyFromLocator(this.locator, body);
  }

  /** Local-space rotation as `(x, y, z)` Euler degrees; exact inverse of {@link setLocalRotation}. */
  async localEuler(): Promise<[number, number, number] | null> {
    const result = (await this.client.execute(
      this.nb(
        "return [_node.getXRotControl().getValue(), _node.getYRotControl().getValue(), _node.getZRotControl().getValue()];",
      ),
    )).value as number[] | null;
    if (result === null) return null;
    return [result[0], result[1], result[2]];
  }

  /** Local-space rotation as `{x, y, z, w}` quaternion (read-only). */
  async localRotation(): Promise<{ x: number; y: number; z: number; w: number } | null> {
    return (await this.client.execute(this.nb("var r = _node.getLocalRot(); return {x: r.x, y: r.y, z: r.z, w: r.w};")))
      .value as { x: number; y: number; z: number; w: number } | null;
  }

  async setLocalRotation(x: number, y: number, z: number): Promise<void> {
    await this.client.execute(
      this.nb(
        `_node.getXRotControl().setValue(${ScriptBuilder.serializeArg(x)}); _node.getYRotControl().setValue(${ScriptBuilder.serializeArg(y)}); _node.getZRotControl().setValue(${ScriptBuilder.serializeArg(z)});`,
      ),
    );
  }

  /** Local-space position as `{x, y, z}` (read-only). */
  async localPosition(): Promise<{ x: number; y: number; z: number } | null> {
    return (await this.client.execute(this.nb("var p = _node.getLocalPos(); return {x: p.x, y: p.y, z: p.z};")))
      .value as { x: number; y: number; z: number } | null;
  }

  /** Rotation order string (e.g. `"XYZ"`), or `null`. */
  async rotationOrder(): Promise<string | null> {
    return (await this.client.execute(this.nb("return _node.getRotationOrder();"))) .value as string | null;
  }

  /** The parent skeleton, or `null`. */
  async getSkeleton(): Promise<import("./skeleton.js").DazSkeleton | null> {
    const name = (await this.client.execute(this.nb("var s = _node.getSkeleton(); return s ? s.getName() : null;")))
      .value as string | null;
    if (name === null) return null;
    const { DazSkeleton } = await import("./skeleton.js");
    return new DazSkeleton(this.client, { value: name, kind: "name" });
  }
}

File: dazpy/_bone.py
from __future__ import annotations

from ._element import DazElement
from ._node import DazNode, NodeIdentifier
from ._script_builder import ScriptBuilder


class DazBone(DazNode):
    """Proxy for a ``DzBone`` (a single joint within a :class:`~dazpy.DazSkeleton`).

    Extends :class:`~dazpy.DazNode` with bone-specific rotation helpers.
    """

    @classmethod
    def _from_locator(cls, client: "DazClient", locator: str, name: str) -> "DazBone":  # noqa: F821
        """Construct a DazBone from a pre-built skeleton-relative locator."""
        bone = object.__new__(cls)
        DazElement.__init__(bone, client, locator)
        object.__setattr__(bone, "_identifier", NodeIdentifier(name))
        return bone

    def _nb(self, body: str) -> str:
        return ScriptBuilder.node_body_from_locator(self._locator, body)

    @property
    def local_euler(self) -> tuple[float, float, float] | None:
        """Local-space rotation as an ``(x, y, z)`` tuple of Euler angles in degrees.

        Reads the same rotation controls written by :meth:`set_local_rotation`,
        so the two are exact inverses.

        Returns:
            ``(x, y, z)`` in degrees, or ``None`` if the bone cannot be found.
        """
        result = self._client.execute(self._nb(
            "return [_node.getXRotControl().getValue(), "
            "_node.getYRotControl().getValue(), "
            "_node.getZRotControl().getValue()];"
        )).value
        if result is None:
            return None
        return (result[0], result[1], result[2])

    @property
    def local_rotation(self) -> dict | None:
        """Local-space rotation as ``{"x", "y", "z", "w"}`` quaternion (read-only)."""
        return self._client.execute(self._nb(
            "var r = _node.getLocalRot(); return {x: r.x, y: r.y, z: r.z, w: r.w};"
        )).value

    def set_local_rotation(self, x: float, y: float, z: float) -> None:
        """Set the bone's local rotation using Euler angles in degrees.

        Args:
            x: Rotation around the local X axis.
            y: Rotation around the local Y axis.
            z: Rotation around the local Z axis.
        """
        self._client.execute(self._nb(
            f"_node.getXRotControl().setValue({float(x)}); "
            f"_node.getYRotControl().setValue({float(y)}); "
            f"_node.getZRotControl().setValue({float(z)});"
        ))

    @property
    def local_position(self) -> dict | None:
        """Local-space position as ``{"x", "y", "z"}`` (read-only)."""
        return self._client.execute(self._nb(
            "var p = _node.getLocalPos(); return {x: p.x, y: p.y, z: p.z};"
        )).value

    @property
    def rotation_order(self) -> str | None:
        """Rotation order string (e.g. ``"XYZ"``), or ``None``."""
        return self._client.execute(self._nb(
            "return _node.getRotationOrder().toString();"
        )).value

    def get_skeleton(self) -> "DazSkeleton | None":  # noqa: F821
        """Return the parent :class:`~dazpy.DazSkeleton`, or ``None``."""
        from ._skeleton import DazSkeleton
        name = self._client.execute(self._nb(
            "var s = _node.getSkeleton(); return s ? s.getName() : null;"
        )).value
        if name is None:
            return None
        return DazSkeleton(self._client, NodeIdentifier(name))
